subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformerInvoker.java

```diff
@@ -112,17 +129,44 @@ private Exception wrapInTransformInvocationException(Class<? extends ArtifactTra
         return new TransformationException(primaryInput.getAbsoluteFile(), transformerType, originalFailure);
     }
 
-    private TransformationIdentity getImmutableTransformationIdentity(File primaryInput, Transformer transformer) {
+    private TransformationIdentity getTransformationIdentity(boolean mutable, File primaryInput, Transformer transformer, TransformationSubject subject) {
+        String initialFileName = subject.getInitialFileName();
+        return mutable ? getMutableTransformationIdentity(primaryInput, transformer, initialFileName) :
+            getImmutableTransformationIdentity(primaryInput, transformer, initialFileName);
+    }
+
+    private TransformationIdentity getImmutableTransformationIdentity(File primaryInput, Transformer transformer, String initialFileName) {
         FileSystemLocationSnapshot snapshot = fileSystemSnapshotter.snapshot(primaryInput);
         return new ImmutableTransformationIdentity(
-            primaryInput.getName(),
+            initialFileName,
             snapshot.getAbsolutePath(),
             snapshot.getHash(),
             transformer.getSecondaryInputHash()
         );
     }
 
-    private Try<ImmutableList<File>> fireTransformListeners(Transformer transformer, TransformationSubject subject, Supplier<Try<ImmutableList<File>>> execution) {
+    private TransformationIdentity getMutableTransformationIdentity(File primaryInput, Transformer transformer, String initialFileName) {
+        return new MutableTransformationIdentity(
+            initialFileName,
+            primaryInput.getAbsolutePath(),
+            transformer.getSecondaryInputHash()
+        );
+    }
+
+    private TransformerExecutionHistoryRepository determineHistoryRepository(Optional<ProjectInternal> producerProject) {
```

Comment:
`Optional` shouldn't be passed around like this. Usually you want to handle the empty case directly where you acquired the `Optional`.

Reply:
I call this method in two places and would need to put in the default handling in two places. I find it better to pass the `Optional` around.

Without using using `Optional` as parameter, I would have
```java
producerProject.map(this::determineHistoryRepository).orElse(gradleUserHomeHistoryRepository)
```

or call `determineProducerProject` twice.

Do you have another suggestion?

Reply:
I am not using `Optional` within the class any more.